{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0ea3ae27",
   "metadata": {},
   "outputs": [],
   "source": [
    "#document_id\tsentence_id\tdoc_start_index\tdoc_end_index\tsentence_entity_start\tsentence_entity_end\ttext_segment\tsemantic_type_id\tentity_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8746bf17",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4c13d13b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>outputs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Create a comma-separated list of all medical t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Create a comma-separated list of all medical t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Create a comma-separated list of all medical t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Create a comma-separated list of all medical t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Create a comma-separated list of all medical t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             outputs\n",
       "0  Create a comma-separated list of all medical t...\n",
       "1  Create a comma-separated list of all medical t...\n",
       "2  Create a comma-separated list of all medical t...\n",
       "3  Create a comma-separated list of all medical t...\n",
       "4  Create a comma-separated list of all medical t..."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_checkpoint_mm_multi_sentences_671\n",
    "df_result_outputs = pd.read_csv('df_checkpoint_sp21_test_sentences_llama_prompt_0929_v1_13b.tsv', sep='\\t', header=0)\n",
    "df_result_outputs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e15c4a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ast import literal_eval\n",
    "#df_result_outputs['outputs'] = df_result_outputs['outputs'].apply(literal_eval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "319b8245",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_outputs = list(df_result_outputs['outputs'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2ccf6a4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Create a comma-separated list of all medical terms in the original language without translations in the medical article below:\\n\"Nonylphenol diethoxylate inhibits apoptosis induced in PC12 cells Nonylphenol and short-chain nonylphenol ethoxylates such as NP2 EO are present in aquatic environment as wastewater contaminants, and their toxic effects on aquatic species have been reported. Apoptosis has been shown to be induced by serum deprivation or copper treatment. To understand the toxicity of nonylphenol diethoxylate, we investigated the effects of NP2 EO on apoptosis induced by serum deprivation and copper by using PC12 cell system.\"\\n\\nHere is the list of medical terms in the original language without translations:\\n\\nNonylphenol, diethoxylate, apoptosis, PC12, serum deprivation, copper, toxicity.',\n",
       " 'Create a comma-separated list of all medical terms in the original language without translations in the medical article below:\\n\"Nonylphenol diethoxylate itself showed no toxicity and recovered cell viability from apoptosis. In addition, nonylphenol diethoxylate decreased DNA fragmentation caused by apoptosis in PC12 cells. This phenomenon was confirmed after treating apoptotic PC12 cells with nonylphenol diethoxylate, whereas the cytochrome c release into the cytosol decreased as compared to that in apoptotic cells not treated with nonylphenol diethoxylate s. Furthermore, Bax contents in apoptotic cells were reduced after exposure to nonylphenol diethoxylate. Thus, nonylphenol diethoxylate has the opposite effect on apoptosis in PC12 cells compared to nonylphenol, which enhances apoptosis induced by serum deprivation.\"\\n\\nHere is the list of medical terms in the original language without translations:\\n\\n* Nonylphenol diethoxylate\\n* apoptosis\\n* PC12 cells\\n* cytochrome c\\n* cytosol\\n* Bax\\n* serum deprivation',\n",
       " 'Create a comma-separated list of all medical terms in the original language without translations in the medical article below:\\n\"The difference in structure of the two compounds is hypothesized to be responsible for this phenomenon. These results indicated that nonylphenol diethoxylate has capability to affect cell differentiation and development and has potentially harmful effect on organisms because of its unexpected impact on apoptosis. © 2015 Wiley Periodicals, Inc. Environ Toxicol 31: 1389-1398, 2016.\"\\n\\nHere is the list of medical terms in the original language without translations:\\n\\n1. structure\\n2. compounds\\n3. phenomenon\\n4. cell differentiation\\n5. development\\n6. apoptosis',\n",
       " 'Create a comma-separated list of all medical terms in the original language without translations in the medical article below:\\n\"EZH2 and ZFX oncogenes in malignant behaviour of parathyroid neoplasms Several studies reported somatic mutations of many genes (MEN1, CTNNB1, CDKIs and others) in parathyroid adenoma, although with different prevalence. Recently, activating mutations of the EZH2 and ZFX oncogenes were identified in benign parathyroid adenoma by whole exome sequencing. The same mutations had been found in blood and ovary malignant tumours.\"\\n\\nHere is the list of medical terms in the original language without translations:\\n\\nEZH2, ZFX, MEN1, CTNNB1, CDKIs',\n",
       " 'Create a comma-separated list of all medical terms in the original language without translations in the medical article below:\\n\"On one hand, this result raised the hypothesis that these oncogenes may play a role in the onset of parathyroid tumour, but it would also suggest they may be involved in malignant, rather benign, parathyroid neoplasm. Our aim was to verify the occurrence of selected mutations of the EZH2 and ZFX genes in an Italian cohort of 23 sporadic parathyroid carcinomas, 12 atypical and 45 typical adenomas. DNA was extracted from paraffin-embedded tissues, PCR amplified and directly sequenced. No mutations were detected in the coding sequence and boundaries of both genes in any of the samples.\"\\n\\nHere is the list of medical terms in the original language without translations:\\n\\n* oncogenes\\n* parathyroid tumour\\n* malignant\\n* benign\\n* parathyroid neoplasm\\n* EZH2\\n* ZFX\\n* DNA\\n* PCR\\n* sequenced',\n",
       " 'Create a comma-separated list of all medical terms in the original language without translations in the medical article below:\\n\"Two polymorphisms of the EZH2 gene were identified with different prevalence: the rs2072407 variant was present in the 30 % of the samples, in keeping with the overall frequency in larger populations, while the rs78589034 variant, located close to the 5\\' end of the exon 16, was detected in only one proband with familial isolated hyperparathyroidism; we investigated the possible outcome on the splicing process. EZH2 and ZFX genes do not seem to have an impact on the onset of most parathyroid tumours, both benign and malignant, though further studies on larger cohorts of different ethnicity are needed.\"\\n\\nHere is the list of medical terms in the original language without translations:\\n\\nEZH2, ZFX, parathyroid tumours, familial isolated hyperparathyroidism.',\n",
       " 'Create a comma-separated list of all medical terms in the original language without translations in the medical article below:\\n\"Patient - Physician Discordance in Global Assessment in Rheumatoid Arthritis: A Systematic Literature Review With Meta-Analysis The integration of the patient in therapeutic decision-making is important in the management of rheumatoid arthritis (RA), but the patient opinion regarding disease status may differ from the physician\\'s opinion. The aim of this study was to assess in the published literature the frequency and drivers of patient - physician discordance in global assessment in RA. A systematic literature review of all articles published up to January 2015 in Medline or Embase, reporting discordance in RA, was conducted by 2 investigators.\"\\n\\nHere is the list of medical terms in the original language without translations:\\n\\n* patient\\n* physician\\n* discordance\\n* global assessment\\n* rheumatoid arthritis\\n* RA\\n* Medline\\n* Embase',\n",
       " 'Create a comma-separated list of all medical terms in the original language without translations in the medical article below:\\n\"Discordance was defined based on the absolute difference of patient global (PGA) and physician global assessments (PhGA) on 0-10-cm scales. The frequency of discordance and its predictors were collected in each study. Frequencies of discordance were pooled by meta-analysis using random effect. In all, 12 studies were selected (i.e., 11,879 patients): weighted mean ± SD age was 55.1 ± 13.9 years, weighted mean ± SD disease duration was 10.4 ± 9.3 years, and 80.7% were women.\"\\n\\nHere is the list of medical terms in the original language without translations:\\n\\n* Discordance\\n* Patient global (PGA)\\n* Physician global assessments (PhGA)\\n* 0-10-cm scales\\n* Disease duration',\n",
       " 'Create a comma-separated list of all medical terms in the original language without translations in the medical article below:\\n\"The value of the difference | PGA - PhGA | defining discordance varied between ≥0.5 cm (n = 2 studies) to ≥3 cm (n = 5 studies); the weighted mean value was 2.7 cm. The pooled percentage of patients with discordance was 43% (95% confidence interval 36%-51%; range 25%-76%). PGA was usually higher than PhGA. The drivers of PGA were pain and functional incapacity, whereas drivers of PhGA were joint counts and acute-phase reactants.\"\\n\\nHere is the list of medical terms in the original language without translations:\\n\\nPGA, PhGA, discordance, cm.',\n",
       " 'Create a comma-separated list of all medical terms in the original language without translations in the medical article below:\\n\"Discordance in global assessment was most frequently defined as a difference of 3 points or more; even with such a stringent definition, up to half the patients were found to be discordant. The long-term consequences of this discordance remain to be determined.\"\\n\\nHere is the list of medical terms in the original language without translations:\\n\\n* discordance\\n* global assessment\\n* long-term consequences']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_outputs[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "34c93648",
   "metadata": {},
   "outputs": [],
   "source": [
    "def strip_llama_prompt(text):\n",
    "    text = text.replace('Create a comma-separated list of all medical terms in the original language without translations in the medical article below:\\n', '')\n",
    "    if '\"\\n\\n' in text:\n",
    "        end_article = text.index('\"\\n\\n')\n",
    "        text = text[end_article+3:]\n",
    "    else:\n",
    "        text = ''\n",
    "    return text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "06aed6d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Here is the list of medical terms in the original language without translations:\\n\\nNonylphenol, diethoxylate, apoptosis, PC12, serum deprivation, copper, toxicity.'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strip_llama_prompt(result_outputs[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1e23b388",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx = 0\n",
    "stripped_results = []\n",
    "for result in result_outputs:    \n",
    "    res = strip_llama_prompt(result)\n",
    "    #print(f'{idx} - {res}')\n",
    "    stripped_results.append(res)\n",
    "    idx+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e474f871",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Here is the list of medical terms in the original language without translations:\\n\\nNonylphenol, diethoxylate, apoptosis, PC12, serum deprivation, copper, toxicity.',\n",
       " 'Here is the list of medical terms in the original language without translations:\\n\\n* Nonylphenol diethoxylate\\n* apoptosis\\n* PC12 cells\\n* cytochrome c\\n* cytosol\\n* Bax\\n* serum deprivation',\n",
       " 'Here is the list of medical terms in the original language without translations:\\n\\n1. structure\\n2. compounds\\n3. phenomenon\\n4. cell differentiation\\n5. development\\n6. apoptosis',\n",
       " 'Here is the list of medical terms in the original language without translations:\\n\\nEZH2, ZFX, MEN1, CTNNB1, CDKIs',\n",
       " 'Here is the list of medical terms in the original language without translations:\\n\\n* oncogenes\\n* parathyroid tumour\\n* malignant\\n* benign\\n* parathyroid neoplasm\\n* EZH2\\n* ZFX\\n* DNA\\n* PCR\\n* sequenced',\n",
       " 'Here is the list of medical terms in the original language without translations:\\n\\nEZH2, ZFX, parathyroid tumours, familial isolated hyperparathyroidism.',\n",
       " 'Here is the list of medical terms in the original language without translations:\\n\\n* patient\\n* physician\\n* discordance\\n* global assessment\\n* rheumatoid arthritis\\n* RA\\n* Medline\\n* Embase',\n",
       " 'Here is the list of medical terms in the original language without translations:\\n\\n* Discordance\\n* Patient global (PGA)\\n* Physician global assessments (PhGA)\\n* 0-10-cm scales\\n* Disease duration',\n",
       " 'Here is the list of medical terms in the original language without translations:\\n\\nPGA, PhGA, discordance, cm.',\n",
       " 'Here is the list of medical terms in the original language without translations:\\n\\n* discordance\\n* global assessment\\n* long-term consequences']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stripped_results[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "690bfdb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_outputs = stripped_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "561bcefe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def unique_occurences(occurences):\n",
    "    # intilize a null list\n",
    "    unique_list = []\n",
    "\n",
    "    # traverse for all elements\n",
    "    for x in occurences:\n",
    "        # check if exists in unique_list or not\n",
    "        if x not in unique_list:\n",
    "            unique_list.append(x)\n",
    "\n",
    "    return unique_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2d1f9097",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "def find_original_span(text, term, abbr=False):    \n",
    "    found_original_text = ''\n",
    "    \n",
    "    search_text = text\n",
    "    search_term = term\n",
    "    if abbr and search_term in search_text:\n",
    "        found_original_text = search_term\n",
    "    else:       \n",
    "        search_text = text.lower()\n",
    "        search_term = term.lower()\n",
    "        if search_term in search_text:\n",
    "            found_original_text = search_term\n",
    "        #else:\n",
    "        #    found_original_text = fuzzy_ngram_match(term, text).lower()\n",
    "\n",
    "    if found_original_text != '':\n",
    "        print(found_original_text)\n",
    "        occurrences_start = [m.start() for m in re.finditer(re.escape(found_original_text), search_text)]\n",
    "        found_occurences = []\n",
    "        for start_idx in occurrences_start:\n",
    "            #start_idx = current_idx + text.lower()[current_idx:].index(found_original_text)\n",
    "            end_idx = start_idx + len(found_original_text)\n",
    "            found_occurences.append({\n",
    "                'start_idx': start_idx,\n",
    "                'end_idx': end_idx,\n",
    "                'term': text[start_idx:end_idx]\n",
    "            })\n",
    "        return found_occurences\n",
    "    else:\n",
    "        return [{\n",
    "            'start_idx': -1,\n",
    "            'end_idx': -1,\n",
    "            'term': 'no-match'\n",
    "        }]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "bd612f5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import re\n",
    "def convert_output_from_comma_list(text, outputs, abbr=False):\n",
    "    outputs = str(outputs)\n",
    "    if ':' in outputs[0:100]:\n",
    "        outputs = outputs[outputs.index(':')+1:]\n",
    "    outputs_replaced = outputs.replace('*', '')\n",
    "    #.replace('\\\\n', ',')\n",
    "    term_lines = [term_line.strip() for term_line in outputs_replaced.split('\\n')]\n",
    "    terms_from_lines = [term_line.split(',') for term_line in term_lines]\n",
    "    terms = [term.strip() for term_line in terms_from_lines for term in term_line]\n",
    "    print(f'terms: {terms}')\n",
    "    original_terms = []\n",
    "    for term in terms:\n",
    "        if len(term.split(' ')) < 10: # longer terms are probably explanations\n",
    "            if term.startswith('1. ') or term.startswith('2. ') or term.startswith('3. ') or term.startswith('4. ') or term.startswith('5. ') or term.startswith('6. ') or term.startswith('7. ') or term.startswith('8. ') or term.startswith('9. '):                \n",
    "                term = term[3:]\n",
    "            \n",
    "            original_term = find_original_span(text, term, abbr)\n",
    "            if len(original_term) > 0 and original_term[0]['start_idx'] > -1:\n",
    "                original_terms.extend(original_term)\n",
    "    \n",
    "    #todo: handle nested entities\n",
    "    return unique_occurences(original_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "2bc4fd8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doc_id</th>\n",
       "      <th>sentence_id</th>\n",
       "      <th>sentence</th>\n",
       "      <th>sentence_doc_start</th>\n",
       "      <th>sentence_doc_end</th>\n",
       "      <th>sentence_starts</th>\n",
       "      <th>tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>25847295</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonylphenol diethoxylate inhibits apoptosis in...</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "      <td>[0, 66, 259, 340]</td>\n",
       "      <td>[{'text': 'Nonylphenol', 'start_char': 0, 'end...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25847295</td>\n",
       "      <td>1</td>\n",
       "      <td>Nonylphenol diethoxylate itself showed no toxi...</td>\n",
       "      <td>514</td>\n",
       "      <td>1215</td>\n",
       "      <td>[0, 96, 197, 540]</td>\n",
       "      <td>[{'text': 'Nonylphenol', 'start_char': 0, 'end...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>25847295</td>\n",
       "      <td>2</td>\n",
       "      <td>The difference in structure of the two compoun...</td>\n",
       "      <td>1216</td>\n",
       "      <td>1598</td>\n",
       "      <td>[0, 104, 315]</td>\n",
       "      <td>[{'text': 'The', 'start_char': 0, 'end_char': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26876532</td>\n",
       "      <td>0</td>\n",
       "      <td>EZH2 and ZFX oncogenes in malignant behaviour ...</td>\n",
       "      <td>0</td>\n",
       "      <td>426</td>\n",
       "      <td>[0, 71, 221, 355]</td>\n",
       "      <td>[{'text': 'EZH2', 'start_char': 0, 'end_char':...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>26876532</td>\n",
       "      <td>1</td>\n",
       "      <td>On one hand, this result raised the hypothesis...</td>\n",
       "      <td>427</td>\n",
       "      <td>1016</td>\n",
       "      <td>[0, 218, 400, 488]</td>\n",
       "      <td>[{'text': 'On', 'start_char': 0, 'end_char': 2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     doc_id  sentence_id                                           sentence   \n",
       "0  25847295            0  Nonylphenol diethoxylate inhibits apoptosis in...  \\\n",
       "1  25847295            1  Nonylphenol diethoxylate itself showed no toxi...   \n",
       "2  25847295            2  The difference in structure of the two compoun...   \n",
       "3  26876532            0  EZH2 and ZFX oncogenes in malignant behaviour ...   \n",
       "4  26876532            1  On one hand, this result raised the hypothesis...   \n",
       "\n",
       "   sentence_doc_start  sentence_doc_end     sentence_starts   \n",
       "0                   0               513   [0, 66, 259, 340]  \\\n",
       "1                 514              1215   [0, 96, 197, 540]   \n",
       "2                1216              1598       [0, 104, 315]   \n",
       "3                   0               426   [0, 71, 221, 355]   \n",
       "4                 427              1016  [0, 218, 400, 488]   \n",
       "\n",
       "                                              tokens  \n",
       "0  [{'text': 'Nonylphenol', 'start_char': 0, 'end...  \n",
       "1  [{'text': 'Nonylphenol', 'start_char': 0, 'end...  \n",
       "2  [{'text': 'The', 'start_char': 0, 'end_char': ...  \n",
       "3  [{'text': 'EZH2', 'start_char': 0, 'end_char':...  \n",
       "4  [{'text': 'On', 'start_char': 0, 'end_char': 2...  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sentences = pd.read_csv('../MedMentions-master/MedMentions-master/st21pv_test_sentences_tokens_combined.tsv', sep='\\t', header=0)\n",
    "df_sentences.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "caa393ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "texts = list(df_sentences['sentence'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "73183a78",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ast import literal_eval\n",
    "df_sentences['tokens'] = df_sentences['tokens'].apply(literal_eval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af9bb6e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_original = []\n",
    "\n",
    "for text, outputs in zip(texts, result_outputs):    \n",
    "    terms = convert_output_from_comma_list(text, outputs)\n",
    "    result_original.append(terms)\n",
    "    \n",
    "    #combined = []\n",
    "    #for output in outputs:\n",
    "    #    combined += convert_output_from_comma_list(text, output)\n",
    "    #result_original.append(combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "921cefe1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2691"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(result_original)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "efcf46cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>outputs</th>\n",
       "      <th>original</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 11, 'term': 'Nony...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 24, 'term': 'Nony...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 18, 'end_idx': 27, 'term': 'str...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 4, 'term': 'EZH2'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 58, 'end_idx': 67, 'term': 'onc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 25, 'end_idx': 29, 'term': 'EZH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 7, 'term': 'Patie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 11, 'term': 'Disc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 30, 'end_idx': 33, 'term': 'PGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 11, 'term': 'Disc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 63, 'end_idx': 81, 'term': 'Ora...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 114, 'end_idx': 127, 'term': 'd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 20, 'term': 'Adve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 10, 'term': 'Mali...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 219, 'end_idx': 222, 'term': 'A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 13, 'end_idx': 29, 'term': 'ult...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 82, 'end_idx': 96, 'term': 'his...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 10, 'end_idx': 18, 'term': 'cof...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 22, 'end_idx': 30, 'term': 'cof...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 15, 'end_idx': 20, 'term': 'H19...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 33, 'end_idx': 56, 'term': 'car...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 117, 'end_idx': 136, 'term': 'a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 9, 'term': 'Hemat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 5, 'end_idx': 13, 'term': 'infr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 19, 'end_idx': 31, 'term': 'ven...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 60, 'end_idx': 72, 'term': 'ven...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 9, 'end_idx': 32, 'term': 'Neph...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 37, 'term': 'Mult...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 17, 'end_idx': 28, 'term': 'nep...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 313, 'end_idx': 354, 'term': 'N...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              outputs   \n",
       "0   Here is the list of medical terms in the origi...  \\\n",
       "1   Here is the list of medical terms in the origi...   \n",
       "2   Here is the list of medical terms in the origi...   \n",
       "3   Here is the list of medical terms in the origi...   \n",
       "4   Here is the list of medical terms in the origi...   \n",
       "5   Here is the list of medical terms in the origi...   \n",
       "6   Here is the list of medical terms in the origi...   \n",
       "7   Here is the list of medical terms in the origi...   \n",
       "8   Here is the list of medical terms in the origi...   \n",
       "9   Here is the list of medical terms in the origi...   \n",
       "10  Here is the list of medical terms in the origi...   \n",
       "11  Here is the list of medical terms in the origi...   \n",
       "12  Here is the list of medical terms in the origi...   \n",
       "13  Here is the list of medical terms in the origi...   \n",
       "14  Here is the list of medical terms in the origi...   \n",
       "15  Here is the list of medical terms in the origi...   \n",
       "16  Here is the list of medical terms in the origi...   \n",
       "17  Here is the list of medical terms in the origi...   \n",
       "18  Here is the list of medical terms in the origi...   \n",
       "19  Here is the list of medical terms in the origi...   \n",
       "20  Here is the list of medical terms in the origi...   \n",
       "21  Here is the list of medical terms in the origi...   \n",
       "22  Here is the list of medical terms in the origi...   \n",
       "23  Here is the list of medical terms in the origi...   \n",
       "24  Here is the list of medical terms in the origi...   \n",
       "25  Here is the list of medical terms in the origi...   \n",
       "26  Here is the list of medical terms in the origi...   \n",
       "27  Here is the list of medical terms in the origi...   \n",
       "28  Here is the list of medical terms in the origi...   \n",
       "29  Here is the list of medical terms in the origi...   \n",
       "\n",
       "                                             original  \n",
       "0   [{'start_idx': 0, 'end_idx': 11, 'term': 'Nony...  \n",
       "1   [{'start_idx': 0, 'end_idx': 24, 'term': 'Nony...  \n",
       "2   [{'start_idx': 18, 'end_idx': 27, 'term': 'str...  \n",
       "3   [{'start_idx': 0, 'end_idx': 4, 'term': 'EZH2'...  \n",
       "4   [{'start_idx': 58, 'end_idx': 67, 'term': 'onc...  \n",
       "5   [{'start_idx': 25, 'end_idx': 29, 'term': 'EZH...  \n",
       "6   [{'start_idx': 0, 'end_idx': 7, 'term': 'Patie...  \n",
       "7   [{'start_idx': 0, 'end_idx': 11, 'term': 'Disc...  \n",
       "8   [{'start_idx': 30, 'end_idx': 33, 'term': 'PGA...  \n",
       "9   [{'start_idx': 0, 'end_idx': 11, 'term': 'Disc...  \n",
       "10  [{'start_idx': 63, 'end_idx': 81, 'term': 'Ora...  \n",
       "11  [{'start_idx': 114, 'end_idx': 127, 'term': 'd...  \n",
       "12  [{'start_idx': 0, 'end_idx': 20, 'term': 'Adve...  \n",
       "13  [{'start_idx': 0, 'end_idx': 10, 'term': 'Mali...  \n",
       "14  [{'start_idx': 219, 'end_idx': 222, 'term': 'A...  \n",
       "15  [{'start_idx': 13, 'end_idx': 29, 'term': 'ult...  \n",
       "16  [{'start_idx': 82, 'end_idx': 96, 'term': 'his...  \n",
       "17  [{'start_idx': 10, 'end_idx': 18, 'term': 'cof...  \n",
       "18  [{'start_idx': 22, 'end_idx': 30, 'term': 'cof...  \n",
       "19  [{'start_idx': 15, 'end_idx': 20, 'term': 'H19...  \n",
       "20  [{'start_idx': 33, 'end_idx': 56, 'term': 'car...  \n",
       "21  [{'start_idx': 117, 'end_idx': 136, 'term': 'a...  \n",
       "22  [{'start_idx': 0, 'end_idx': 9, 'term': 'Hemat...  \n",
       "23  [{'start_idx': 5, 'end_idx': 13, 'term': 'infr...  \n",
       "24  [{'start_idx': 19, 'end_idx': 31, 'term': 'ven...  \n",
       "25  [{'start_idx': 60, 'end_idx': 72, 'term': 'ven...  \n",
       "26  [{'start_idx': 9, 'end_idx': 32, 'term': 'Neph...  \n",
       "27  [{'start_idx': 0, 'end_idx': 37, 'term': 'Mult...  \n",
       "28  [{'start_idx': 17, 'end_idx': 28, 'term': 'nep...  \n",
       "29  [{'start_idx': 313, 'end_idx': 354, 'term': 'N...  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_small_results = pd.DataFrame({'outputs': result_outputs, 'original': result_original}) #, 'medical_term': result_medical_terms})\n",
    "df_small_results.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "410595a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_outputs = df_small_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "d923b3bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_subset = df_sentences#[0:1221] #.tail(600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "7c426952",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_outputs['sentence'] = list(df_subset['sentence'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "e6941ec0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>outputs</th>\n",
       "      <th>original</th>\n",
       "      <th>sentence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 11, 'term': 'Nony...</td>\n",
       "      <td>Nonylphenol diethoxylate inhibits apoptosis in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 24, 'term': 'Nony...</td>\n",
       "      <td>Nonylphenol diethoxylate itself showed no toxi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 18, 'end_idx': 27, 'term': 'str...</td>\n",
       "      <td>The difference in structure of the two compoun...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 4, 'term': 'EZH2'...</td>\n",
       "      <td>EZH2 and ZFX oncogenes in malignant behaviour ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 58, 'end_idx': 67, 'term': 'onc...</td>\n",
       "      <td>On one hand, this result raised the hypothesis...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             outputs   \n",
       "0  Here is the list of medical terms in the origi...  \\\n",
       "1  Here is the list of medical terms in the origi...   \n",
       "2  Here is the list of medical terms in the origi...   \n",
       "3  Here is the list of medical terms in the origi...   \n",
       "4  Here is the list of medical terms in the origi...   \n",
       "\n",
       "                                            original   \n",
       "0  [{'start_idx': 0, 'end_idx': 11, 'term': 'Nony...  \\\n",
       "1  [{'start_idx': 0, 'end_idx': 24, 'term': 'Nony...   \n",
       "2  [{'start_idx': 18, 'end_idx': 27, 'term': 'str...   \n",
       "3  [{'start_idx': 0, 'end_idx': 4, 'term': 'EZH2'...   \n",
       "4  [{'start_idx': 58, 'end_idx': 67, 'term': 'onc...   \n",
       "\n",
       "                                            sentence  \n",
       "0  Nonylphenol diethoxylate inhibits apoptosis in...  \n",
       "1  Nonylphenol diethoxylate itself showed no toxi...  \n",
       "2  The difference in structure of the two compoun...  \n",
       "3  EZH2 and ZFX oncogenes in malignant behaviour ...  \n",
       "4  On one hand, this result raised the hypothesis...  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_outputs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "e55e1387",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_outputs['entities'] = df_outputs['original']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "7250a7ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for index, row in df_outputs.iterrows():\n",
    "#    text = row['sentence']\n",
    "#    outputs = row['outputs']\n",
    "#    entities = convert_output_from_comma_list(text, outputs)\n",
    "#    df_outputs.at[index, 'entities'] = entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "992727c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#document_id\tsentence_id\tdoc_start_index\tdoc_end_index\t\n",
    "#sentence_entity_start\tsentence_entity_end\ttext_segment\tsemantic_type_id\tentity_id\n",
    "#doc_id\tsentence_id\tsentence\tsentence_doc_start\tsentence_doc_end\ttokens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "ada0cde0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_outputs['document_id'] = list(df_subset['doc_id'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "cb99db97",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_outputs['sentence_id'] = list(df_subset['sentence_id'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "ddbcb507",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_outputs['doc_start_index'] = list(df_subset['sentence_doc_start'].values)\n",
    "df_outputs['doc_end_index'] = list(df_subset['sentence_doc_end'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "82e2df27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>outputs</th>\n",
       "      <th>original</th>\n",
       "      <th>sentence</th>\n",
       "      <th>entities</th>\n",
       "      <th>document_id</th>\n",
       "      <th>sentence_id</th>\n",
       "      <th>doc_start_index</th>\n",
       "      <th>doc_end_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 11, 'term': 'Nony...</td>\n",
       "      <td>Nonylphenol diethoxylate inhibits apoptosis in...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 11, 'term': 'Nony...</td>\n",
       "      <td>25847295</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 24, 'term': 'Nony...</td>\n",
       "      <td>Nonylphenol diethoxylate itself showed no toxi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 24, 'term': 'Nony...</td>\n",
       "      <td>25847295</td>\n",
       "      <td>1</td>\n",
       "      <td>514</td>\n",
       "      <td>1215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 18, 'end_idx': 27, 'term': 'str...</td>\n",
       "      <td>The difference in structure of the two compoun...</td>\n",
       "      <td>[{'start_idx': 18, 'end_idx': 27, 'term': 'str...</td>\n",
       "      <td>25847295</td>\n",
       "      <td>2</td>\n",
       "      <td>1216</td>\n",
       "      <td>1598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 4, 'term': 'EZH2'...</td>\n",
       "      <td>EZH2 and ZFX oncogenes in malignant behaviour ...</td>\n",
       "      <td>[{'start_idx': 0, 'end_idx': 4, 'term': 'EZH2'...</td>\n",
       "      <td>26876532</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Here is the list of medical terms in the origi...</td>\n",
       "      <td>[{'start_idx': 58, 'end_idx': 67, 'term': 'onc...</td>\n",
       "      <td>On one hand, this result raised the hypothesis...</td>\n",
       "      <td>[{'start_idx': 58, 'end_idx': 67, 'term': 'onc...</td>\n",
       "      <td>26876532</td>\n",
       "      <td>1</td>\n",
       "      <td>427</td>\n",
       "      <td>1016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             outputs   \n",
       "0  Here is the list of medical terms in the origi...  \\\n",
       "1  Here is the list of medical terms in the origi...   \n",
       "2  Here is the list of medical terms in the origi...   \n",
       "3  Here is the list of medical terms in the origi...   \n",
       "4  Here is the list of medical terms in the origi...   \n",
       "\n",
       "                                            original   \n",
       "0  [{'start_idx': 0, 'end_idx': 11, 'term': 'Nony...  \\\n",
       "1  [{'start_idx': 0, 'end_idx': 24, 'term': 'Nony...   \n",
       "2  [{'start_idx': 18, 'end_idx': 27, 'term': 'str...   \n",
       "3  [{'start_idx': 0, 'end_idx': 4, 'term': 'EZH2'...   \n",
       "4  [{'start_idx': 58, 'end_idx': 67, 'term': 'onc...   \n",
       "\n",
       "                                            sentence   \n",
       "0  Nonylphenol diethoxylate inhibits apoptosis in...  \\\n",
       "1  Nonylphenol diethoxylate itself showed no toxi...   \n",
       "2  The difference in structure of the two compoun...   \n",
       "3  EZH2 and ZFX oncogenes in malignant behaviour ...   \n",
       "4  On one hand, this result raised the hypothesis...   \n",
       "\n",
       "                                            entities  document_id   \n",
       "0  [{'start_idx': 0, 'end_idx': 11, 'term': 'Nony...     25847295  \\\n",
       "1  [{'start_idx': 0, 'end_idx': 24, 'term': 'Nony...     25847295   \n",
       "2  [{'start_idx': 18, 'end_idx': 27, 'term': 'str...     25847295   \n",
       "3  [{'start_idx': 0, 'end_idx': 4, 'term': 'EZH2'...     26876532   \n",
       "4  [{'start_idx': 58, 'end_idx': 67, 'term': 'onc...     26876532   \n",
       "\n",
       "   sentence_id  doc_start_index  doc_end_index  \n",
       "0            0                0            513  \n",
       "1            1              514           1215  \n",
       "2            2             1216           1598  \n",
       "3            0                0            426  \n",
       "4            1              427           1016  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_outputs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6e252bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_outputs['original'] = df_outputs['original'].apply(literal_eval)\n",
    "#df_outputs['entities'] = df_outputs['entities'].apply(literal_eval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "6c457252",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_entities = []\n",
    "for index, row in df_outputs.iterrows():\n",
    "    for entity in row['entities']:\n",
    "        if entity['start_idx'] > 0:\n",
    "            result_entities.append({\n",
    "                \"document_id\": row[\"document_id\"],\n",
    "                \"sentence_id\": row[\"sentence_id\"],\n",
    "                \"doc_start_index\": row[\"doc_start_index\"],\n",
    "                \"doc_end_index\": row[\"doc_end_index\"],\n",
    "                \"sentence_entity_start\": entity[\"start_idx\"],\n",
    "                \"sentence_entity_end\": entity[\"end_idx\"],\n",
    "                \"text_segment\": entity[\"term\"]\n",
    "            })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "190cf0d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>document_id</th>\n",
       "      <th>sentence_id</th>\n",
       "      <th>doc_start_index</th>\n",
       "      <th>doc_end_index</th>\n",
       "      <th>sentence_entity_start</th>\n",
       "      <th>sentence_entity_end</th>\n",
       "      <th>text_segment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>25847295</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "      <td>66</td>\n",
       "      <td>77</td>\n",
       "      <td>Nonylphenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25847295</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "      <td>94</td>\n",
       "      <td>105</td>\n",
       "      <td>nonylphenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>25847295</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "      <td>370</td>\n",
       "      <td>381</td>\n",
       "      <td>nonylphenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>25847295</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "      <td>12</td>\n",
       "      <td>24</td>\n",
       "      <td>diethoxylate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>25847295</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "      <td>382</td>\n",
       "      <td>394</td>\n",
       "      <td>diethoxylate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   document_id  sentence_id  doc_start_index  doc_end_index   \n",
       "0     25847295            0                0            513  \\\n",
       "1     25847295            0                0            513   \n",
       "2     25847295            0                0            513   \n",
       "3     25847295            0                0            513   \n",
       "4     25847295            0                0            513   \n",
       "\n",
       "   sentence_entity_start  sentence_entity_end  text_segment  \n",
       "0                     66                   77   Nonylphenol  \n",
       "1                     94                  105   nonylphenol  \n",
       "2                    370                  381   nonylphenol  \n",
       "3                     12                   24  diethoxylate  \n",
       "4                    382                  394  diethoxylate  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_result_entities = pd.DataFrame(result_entities)\n",
    "df_result_entities.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc208613",
   "metadata": {},
   "outputs": [],
   "source": [
    "# todo: remove overlaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "d1ef5f9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_result_entities.to_csv('st21pv_test_0928v1llama13b_entities_nofuzzy.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27f6c814",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
